### PR TITLE
api: Summarize EquivalentAddressGroup toString

### LINKED ISSUE #12593: Summarize EquivalentAddressGroup.toString() output for large address lists
Hello maintainers,

I’ve been interested in gRPC and this project, and I’d like to start contributing as an open source contributor.

### Is your feature request related to a problem?
https://github.com/grpc/grpc-java/blob/6422092e3aebb5b29989d4eb0af27b0b46f007a4/api/src/main/java/io/grpc/EquivalentAddressGroup.java#L114-L118

Yes.  
Currently, `EquivalentAddressGroup.toString()` prints the full list of addresses.  
When a name resolver returns a large number of endpoints (for example, large DNS responses or xDS endpoint lists), this can result in extremely long log lines. This negatively impacts log readability and makes it harder to identify useful information during debugging or operational analysis.

### Describe the solution you'd like

I would like to improve `EquivalentAddressGroup.toString()` so that the output is summarized when the address list is very large, while preserving the current behavior for small lists.

The general idea is:
- Keep the existing full output when the number of addresses is below a certain threshold.
- When the list exceeds that threshold, return a summarized representation
  (for example, showing the first N addresses followed by a “… X more” indicator).
- Preserve the attributes output as-is so that debugging context is not lost.

This change would be limited to the string representation only and would not affect any functional behavior.

Before moving forward, I wanted to check whether this approach sounds reasonable and whether it would be acceptable for me to implement this change and open a PR.

### Describe alternatives you've considered

An alternative would be to leave the output unchanged and rely on external log truncation or logging configuration. However, this does not fully address readability issues and can still make it difficult to understand logs when working with large endpoint sets.

Another option would be to truncate the address list entirely after a fixed number of entries, but that could remove useful context. A summarized format that indicates how many addresses were omitted seems to provide a better balance between readability and debuggability.

### Additional context

There is already a TODO comment in `EquivalentAddressGroup.toString()` noting that the return value should be summarized when the address list is very large. This proposal is intended to address that TODO.

The change is intentionally low risk, scoped only to `toString()`, and aimed at improving observability in high-endpoint environments. I would appreciate any guidance on preferred output format or threshold values.

Thank you very much for your time and consideration.

### PR DESCRIPTION
**Pull request by stagegrowth**
### Background
EquivalentAddressGroup.toString() currently prints the full address list, which can become extremely large and noisy for big address sets. This makes logs and test output hard to read and can obscure useful information.

### Changes
This PR adds a hard cap to the number of addresses rendered in toString() and summarizes the remainder (e.g., “... N more”). It also adds focused unit tests to validate the large-list and boundary behaviors, and refactors related CallOptionsTest assertions using a small Truth subject for clearer equivalence checks.

### Purpose
Keep toString() output readable even with very large address lists
Ensure the summarization behavior is covered by unit tests
Improve clarity of CallOptions equivalence assertions in tests

### Note
The summary threshold is defined as a private constant and is only used for output formatting. It does not change the underlying address data or behavior—only the string representation.

Fixes https://github.com/grpc/grpc-java/issues/12593

### CODE REVIEW

File: api/src/test/java/io/grpc/CallOptionsTest.java
Comment (kannanjgithub) [inline api/src/test/java/io/grpc/CallOptionsTest.java:43]: The changes in this test seem  unrelated to the PR description. Can you move it to a separate PR?

Comment (stagegrowth) [reply api/src/test/java/io/grpc/CallOptionsTest.java:43]: Thanks for the feedback. I’ve reverted the CallOptionsTest refactor in this PR since it’s unrelated to the change. I can follow up with a separate PR if we still want that refactor.

File: api/src/main/java/io/grpc/EquivalentAddressGroup.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
  */
 @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1770")
 public final class EquivalentAddressGroup {
+  private static final int MAX_ADDRESSES_TO_STRING = 100;
```
Comment (kannanjgithub) [inline api/src/main/java/io/grpc/EquivalentAddressGroup.java]: Just make it package-private and access it directly from the tests without the use of reflection.

**Review by ejona86 — COMMENTED** [review]
I don't think this change will actually do anything in practice.

It is very rare to have more than 2 addresses in an EAG. When there are many addresses, that is a list of EAGs. Did you actually test this improves the situation for whatever you were seeing?

File: api/src/main/java/io/grpc/EquivalentAddressGroup.java (hunk 1 above)
Comment (stagegrowth) [reply api/src/main/java/io/grpc/EquivalentAddressGroup.java]: @kannanjgithub Thanks for the suggestion. I’ve made MAX_ADDRESSES_TO_STRING package-private and updated the tests to access it directly, removing the reflection helper.

**Comment by stagegrowth** [comment]
> I don't think this change will actually do anything in practice.
> 
> It is very rare to have more than 2 addresses in an EAG. When there are many addresses, that is a list of EAGs. Did you actually test this improves the situation for whatever you were seeing?


Hi, thanks for the thoughtful feedback.

I agree that in the xDS path it’s uncommon to have more than two addresses in a single EquivalentAddressGroup, and that large endpoint sets are usually represented as a list of EAGs.

My understanding, though, is that EquivalentAddressGroup is used more broadly than just xDS. For example, in the grpclb path or in some DNS- or proxy-based NameResolver implementations, a single EAG may contain multiple addresses, in which case toString() could still produce very large log lines. Please let me know if this understanding is off or if there are important constraints I’m missing.

With that in mind, this change is not meant to target xDS specifically, but rather to explore a small improvement to EAG logging robustness in more general or atypical cases.

To help make this concrete, I added a small test that constructs an EAG with a larger number of addresses, mainly to make the before/after toString() behavior easier to observe.

Thanks again for the review and the helpful questions.

**Comment by ejona86** [comment]
> For example, in the grpclb path or in some DNS- or proxy-based NameResolver implementations, a single EAG may contain multiple addresses, in which case toString() could still produce very large log lines.

I was agreeing it could contain multiple, but like "two." The point of multiple addresses in an EAG is for addresses to the same machine, like IPv4 and IPv6. I'd love people to use it for local-link IPv6 addresses, and potentially multiple public IPv6 addresses, but we're still talking relatively few.

Back [when the TODO was added](https://github.com/grpc/grpc-java/commit/35f0d15291caaa5f21cd986429a17e7f986179b0), the pick first policy would combine all the EAGs into a single large EAG. But [we stopped doing that](https://github.com/grpc/grpc-java/commit/e3ff1ade078fdbb566de5944048eae76b0c87add) later that year. So I think the TODO should just be deleted.

**Comment by stagegrowth** [comment]
> > For example, in the grpclb path or in some DNS- or proxy-based NameResolver implementations, a single EAG may contain multiple addresses, in which case toString() could still produce very large log lines.
> 
> I was agreeing it could contain multiple, but like "two." The point of multiple addresses in an EAG is for addresses to the same machine, like IPv4 and IPv6. I'd love people to use it for local-link IPv6 addresses, and potentially multiple public IPv6 addresses, but we're still talking relatively few.
> 
> Back [when the TODO was added](https://github.com/grpc/grpc-java/commit/35f0d15291caaa5f21cd986429a17e7f986179b0), the pick first policy would combine all the EAGs into a single large EAG. But [we stopped doing that](https://github.com/grpc/grpc-java/commit/e3ff1ade078fdbb566de5944048eae76b0c87add) later that year. So I think the TODO should just be deleted.

@ejona86 

Thanks for the detailed explanation — that clarifies things a lot.

I understand now that the intended semantics of EquivalentAddressGroup are a small number of addresses referring to the same machine (e.g., IPv4/IPv6), and that large address lists in a single EAG are no longer expected in the current design.

Given that the original motivation for the TODO was tied to the old pick_first behavior of merging many EAGs into one — which no longer applies — I agree that adding summarization logic to toString() isn’t the right direction here.

I’ll update the PR accordingly by removing the summarization change (and the associated test) and instead just deleting the outdated TODO comment.

Thanks again for taking the time to explain the background and for the guidance.

**Comment by stagegrowth** [comment]
> > For example, in the grpclb path or in some DNS- or proxy-based NameResolver implementations, a single EAG may contain multiple addresses, in which case toString() could still produce very large log lines.
> 
> I was agreeing it could contain multiple, but like "two." The point of multiple addresses in an EAG is for addresses to the same machine, like IPv4 and IPv6. I'd love people to use it for local-link IPv6 addresses, and potentially multiple public IPv6 addresses, but we're still talking relatively few.
> 
> Back [when the TODO was added](https://github.com/grpc/grpc-java/commit/35f0d15291caaa5f21cd986429a17e7f986179b0), the pick first policy would combine all the EAGs into a single large EAG. But [we stopped doing that](https://github.com/grpc/grpc-java/commit/e3ff1ade078fdbb566de5944048eae76b0c87add) later that year. So I think the TODO should just be deleted.

@ejona86 @kannanjgithub Thanks again for the review. I opened a follow-up PR that removes the obsolete TODO in EquivalentAddressGroup#toString() based on your feedback: #12627
